### DISCUSSION TITLE
Performance degradation during load

### DISCUSSION THREAD

**Discussion by vitaliishandra** [JanusGraph/janusgraph#3566] [Q&A]
We have the operation on backend that performs few janusgraph queries. The total execution time of operation is ~100ms average. We use indexes on vertexes and not full graph scans. 
When we start to perform multiple operations (100-200) in paralell the performance started to degradate (5000-9000ms).
We tried different configurations without any results.
All metrics and configurations are below. Are there any ideas what can be improved?

- Version: 0.6.2
- Storage Backend: cql (tested on Cassandra 3.10.11 and ScyllaDB) 
- Mixed Index Backend: none

Configuration:
3 pods of Janusgraph with following deployment configuration:
         env:
            - name: JAVA_OPTIONS
              value: >-
                -Xms256m -Xmx6144m -Dcom.sun.management.jmxremote=true
                -Dcom.sun.management.jmxremote.authenticate=false
                -Dcom.sun.management.jmxremote.ssl=false
                -Dcom.sun.management.jmxremote.local.only=false
                -Dcom.sun.management.jmxremote.port=9999
                -Dcom.sun.management.jmxremote.rmi.port=9999
            - name: gremlinserver.threadPoolWorker
              value: '16'
            - name: gremlinserver.gremlinPool
              value: '32'
            - name: gremlinserver.writeBufferHighWaterMark
              value: '6553600'
            - name: gremlinserver.writeBufferLowWaterMark
              value: '627680'
            - name: gremlinserver.maxContentLength
              value: '365536'
            - name: gremlinserver.resultIterationBatchSize
              value: '128'
            - name: gremlinserver.evaluationTimeout
              value: '1800000'
            - name: janusgraph.storage.cql.local-datacenter
              value: northcentralus
            - name: janusgraph.cassandra.keyspace
              value: janusgraph
            - name: JANUS_PROPS_TEMPLATE
              value: cql
            - name: janusgraph.storage.hostname
              value: *****
            - name: janusgraph.ids.block-size
              value: '100000'
            - name: janusgraph.storage.cql.read-consistency-level
              value: QUORUM
            - name: janusgraph.storage.cql.write-consistency-level
              value: QUORUM
            - name: janusgraph.storage.cql.replication-factor
              value: '3'
            - name: janusgraph.cache.db-cache
              value: 'false'
            - name: janusgraph.query.batch
              value: 'true'
            - name: janusgraph.query.batch-property-prefetch
              value: 'true'
            - name: janusgraph.query.smart-limit
              value: 'false'
            - name: janusgraph.metrics.enabled
              value: 'true'
            - name: janusgraph.metrics.prefix
              value: janusgraph
            - name: janusgraph.metrics.jmx.enabled
              value: 'true'
            - name: janusgraph.metrics.jmx.domain
              value: janusgraph
            - name: janusgraph.storage.batch-loading
              value: 'false'
            - name: janusgraph.storage.page-size
              value: '500'
            - name: janusgraph.storage.cql.batch-statement-size
              value: '100'
          resources:
            limits:
              cpu: 850m
              memory: 6096Mi
            requests:
              cpu: 650m
              memory: 5096Mi

CQL janusgraph.graphindex latency during load:
![image](https://user-images.githubusercontent.com/56965286/217788854-a89d8384-3063-486f-a640-9ce1e1f60469.png)

CQL janusgraph.edgestore latency during load:
![image](https://user-images.githubusercontent.com/56965286/217789074-75f8b40e-e7e0-45c6-96db-dbaf5395f0fb.png)

Grmelin dashboard:
![image](https://user-images.githubusercontent.com/56965286/217794220-9a92a8b9-e052-4aba-a857-66e717e34082.png)

Janusgraph dashabords:
![image](https://user-images.githubusercontent.com/56965286/217794728-9bbddc1c-68ec-4bff-ad28-89aeceb9a143.png)

**Comment by porunov** [comment]
By default configuration isn't tuned and the throughput is limited. I think it's better to tune the default configuration for new JanusGraph users, but for now I would suggest taking a look at the following configuration options.
1) I would check out specific `cql` configuration options. Perhaps `storage.cql.local-max-connections-per-host` or `storage.cql.max-requests-per-connection` should be tuned. 
2) If you find some CQL driver configuration options [here](https://docs.datastax.com/en/developer/java-driver/4.14/manual/core/configuration/reference/) which are not provided directly in JanusGraph you can always relay on [storage.cql.internal](https://docs.janusgraph.org/configs/configuration-reference/#storagecqlinternal) to pass those configurations directly into the driver.
3) `storage.cql.executor-service.enabled` - I would use it as `false` because for now this thread pool isn't useful, but is used to limit the amount of parallel cql requests. This pool is `enabled` for historical reason, but it may be useful later when we switch to async JanusGraph queries execution. As for now I would recommend disabling it in most cases. This pool shouldn't limit performance much but it may take up additional resources.
4) Perhaps `storage.parallel-backend-executor-service` should be tuned for your queries. Be default it's fixed with the size of `number of processors multiplied by 2` which might be too small for many use-cases. Try either increasing `storage.parallel-backend-executor-service.core-pool-size` to something meaningful in your case or even switch the implementation to some scalable executor pool. For example, you can reuse `org.elasticsearch.common.util.concurrent.EsThreadPoolExecutor` by wrapping it into your own implementation for JanusGraph (notice that it should have the same package path. I.e. you can create `org.elasticsearch.common.util.concurrent.EsThreadPoolExecutorWithJanusGraphSettings`). Nevertheless, it's not necessary if your load is constant and you can simply increase `storage.parallel-backend-executor-service.core-pool-size` (try simply multiply initial value to the amount of parallel requests. i.e. should be `2000` or something in your case).
5) Generally check configurations which might influence performance for your usecase https://docs.janusgraph.org/configs/configuration-reference/ .

I don't know how heavy operations you are performing but usually it's not the bottleneck of Scylla nodes. Nevertheless you should check ScyllaDB metrics as well to verify there is no any limitations on that side. Usually if you have good network capping to handle all load it shouldn't be a problem with network throughput either (nevertheless it's always better to check just in case).
I would assume that most likely your current limitation is in JanusGraph nodes. It looks like your queries are fighting for the same resources but the task is to find what they are fighting for (it's either machine resources like processor time or queries execution resources like thread pool limits or else).

**Reply by vitaliishandra** [reply]
Thank you! We are trying the reccomended steps. Actually the issue is reproduced even on such simple query:
g.V().has('uid', 'Test:1').out(). The 'uid' property value is indexed and is unique all over the graph. 
When I run this query it tooks <100ms. However when i run it in paralell (20 paralell queries) is started to degradate to 4000-6000ms

**Reply by porunov** [reply]
That's most likely resources problem. Most likely other queries simply wait until previously fired queries are executed. I would bet it on `storage.parallel-backend-executor-service.core-pool-size`, but can be anything described above. Let me know what you find.

cc @li-boxuan @cdegroc were you able to replace that thread pool with async approach by any chance?

**Reply by vitaliishandra** [reply]
@porunov I have just tried to set it to 1000/2000 but result is the same. The configuration is correctly set I suppose as during the strat of Jnausgraph I see updated values in logs:
![image](https://user-images.githubusercontent.com/56965286/218076468-5fb55510-35d1-4c1a-9365-7637de704bb4.png)

janusgraph.storage.cql.executor-service.enabled=false
janusgraph.storage.parallel-backend-ops=true
Can these or other configs impact on storage.parallel-backend-executor-service.core-pool-size ?

**Reply by porunov** [reply]
> @porunov I have just tried to set it to 1000/2000 but result is the same. The configuration is correctly set I suppose as during the strat of Jnausgraph I see updated values in logs: ![image](https://user-images.githubusercontent.com/56965286/218076468-5fb55510-35d1-4c1a-9365-7637de704bb4.png)
> 
> janusgraph.storage.cql.executor-service.enabled=false janusgraph.storage.parallel-backend-ops=true Can these or other configs impact on storage.parallel-backend-executor-service.core-pool-size ?

No, they don't affect the pool size.
I think you will need to enable profiling in DataDog to have a more clear picture where queries spend most of the time. 
I also suggest trying to investigate CQL and JanusGraph configuration options (as per `2` and `5` above).
We also use parallel JanusGraph OLTP queries, but in our case the performance is constant for them. Can't say exactly why you observe performance degradation, but would be great you share the reason when you find it.
I want to mention that we are using unreleased JanusGraph version (i.e. 1.0.0 from `master`), but I don't think there was any bugs in 0.6.x versions which would cause performance degradation for parallel queries.

**Reply by li-boxuan** [reply]
I don't think it has much to do with the executor pool since your throughput is very low:

> However when i run it in paralell (20 paralell queries)

And yeah you might have to do some profiling to figure out what happens.

**Reply by vitaliishandra** [reply]
@porunov @li-boxuan  thank you for you reccomendations. We have experimented with different configurations without any success. Maybe you have any other ideas what can be wrong? Here are additional details:
We run multiple queries in paralell. In the same time we connected to Janusgraph pod and run simple query g.V().has('uid', 'Flow:1012309').out().profile() that return 80 results. We are waiting ~30sec to receive the result. However the profile shows that everything is ok and it took 3-4ms:
![image](https://user-images.githubusercontent.com/56965286/218785147-9f09ebdf-1048-4593-a03f-f2de8a60838d.png)

**Reply by porunov** [reply]
30 seconds to return 80 elements doesn't sound right to me. I guess in your case the problem here could be that connection between your application / gremlin console and JanusGraph is too slow. You are probably measuring performance when you received data back into your application, but profile measures how much time it takes to retrieve data from storage backend / index backend and explains how much time it spend on each step.
As such, try to check your serialization config (use GraphBinary if possible), check latency between your app and JanusGraph.
That said, still, I don't imagine how small throughput and high latency should be to return 80 vertices without any properties (because `.out()` will return basically vertex ids with no properties). Thus, it could be that the issue is somewhere else.
The parallelization problem also doesn't make sense to me. If you applied the suggested configuration above it should work normally and you shouldn't see performance degradation if you run 1 or 100 queries like the above one.

The one thing comes to my mind: Are you using Gremlin Console to measure performance by any chance? If yes than it's better not to. I believe the default Gremlin Console thread pool is `1` which means all your queries will be processed one by one (and not in parallel). 

That said, if you are using direct JanusGraph connection to measure performance then I would do the next:
1. Check how much time JanusGraph spends to compute the result (in the above profiling it looks like 3-4 ms).
2. Enable profiling on JanusGraph side and investigate how much time you spend before you send the result to your network (i.e. how much time it takes for serialization or maybe something else).
3. Enable profiling on your application side and check how much time it takes to deserialize the result.
4. Check latency and network throughput.

**Reply by vitaliishandra** [reply]
@porunov  we are trying to perform test on separate VM with Janusgraph+Scylla. We would like to make sure we setup configuration correctly. We use cql and janusgraph-cql.properties file. We set some configuration there. Is it correct to override LOCAL configuration setting via this file (like storage.cql.executor-service.enabled) or it should be overrided in some other way? Is it possible to make sure it is set correctly as management doesn't show it?
![image](https://user-images.githubusercontent.com/56965286/219704798-e286e2f2-6aeb-4ac4-981a-962318ebf4b4.png)

![image](https://user-images.githubusercontent.com/56965286/219704051-1a38ee3a-0417-4380-adb0-3dc9d94ffd34.png)

**Reply by hadoopmarc** [reply]
Overriding LOCAL properties in the properties file is OK, see [here](https://docs.janusgraph.org/configs/configuration-reference/#mutability-levels) for more explanation.
Inspecting actual config properties of the current instance goes like this:

```
gremlin> graph.configuration().getKeys()
==>gremlin.graph
==>storage.backend
==>storage.hostname
==>storage.cql.keyspace
==>storage.cql.local-datacenter
==>cache.db-cache
==>cache.db-cache-clean-wait
==>cache.db-cache-time
==>cache.db-cache-size
gremlin> graph.configuration().getProperty('storage.backend')
==>cql
```

**Comment by hadoopmarc** [comment]
It strikes me that the reads on the graphindex take much longer than those on the edgestore. Are you sure that the indexed property values have a high cardinatlity? See the section "caveat" in this blog: https://li-boxuan.medium.com/janusgraph-deep-dive-part-2-demystify-indexing-d26e71edb386

**Reply by vitaliishandra** [reply]
Yes. We have indexes on few props with high cardinality. We use just composite indexes. It executes very fast without load. However when we produce the paralell connections it starts to degradate.

**Reply by li-boxuan** [reply]
> It strikes me that the reads on the graphindex take much longer than those on the edgestore

Cannot agree more. I would guess it's a coincidence because apparently @vitaliishandra  doesn't have too much traffic as of now. I would say it looks like your Cassandra/Scylla instances don't have enough machine resources. You may need to pull more metrics of them.
